File: C++/sub_sequence.cpp
/**
 * @File: sub_sequence.cpp
 * @Brief: 判断子序列
 * @Link: https://leetcode.cn/problems/is-subsequence/
 * @Author: Wei Li (weili_yzzcq@163.com)
 * @Date: 2022-09-24
*/

#include <iostream>
#include <string>
#include <map>


class Solution 
{
public:
    bool isSubsequence(std::string s, std::string t) 
    {
        int first_ptr_s = 0, second_ptr_t = 0; /* 双指针思想 */
        while (first_ptr_s < s.length() && second_ptr_t < t.length())
        {
            if (s[first_ptr_s] == t[second_ptr_t]) /* 贪心思想 */
            {
                ++first_ptr_s;
            }
            ++second_ptr_t;
        }

        return first_ptr_s == s.length(); /* s 是否匹配结束 */
    }
};


// -----------------------------
int main(int argc, char **argv)
{
    std::string str_1 = "axc";
    std::string str_2 = "ahbgdc";
    // std::string str_1 = "abc";
    // std::string str_2 = "ahbgdc";

    Solution solution;
    auto sub_sequence = solution.isSubsequence(str_1, str_2);
    std::cout << "The solution of this problem is " << std::boolalpha << sub_sequence << std::endl;

    return 0;
}

File: Python/sub_sequence.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: sub_sequence.py
@Function: 判断子序列
@Link: https://leetcode.cn/problems/is-subsequence/
@Python Version: 3.9.7
@Author: Wei Li (weili_yzzcq@163.com)
@Date: 2022-09-24
"""


class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        ptr_s, ptr_t = 0, 0 # 双指针思想
        while (ptr_s < len(s)) and (ptr_t < len(t)):
            if s[ptr_s] == t[ptr_t]: # 贪心思想
                ptr_s +=1
            ptr_t += 1

        return ptr_s == len(s) # s 是否匹配完全


# ----------------------------------
if __name__ == "__main__":
    # str_1 = "axc"
    # str_2 = "ahbgdc"
    str_1 = "abc"
    str_2 = "ahbgdc"

    solution = Solution()
    sub_sequence = solution.isSubsequence(str_1, str_2)
    print(f"The solution of this problem is : {sub_sequence}")
